cpmpy/transformations/int2bool.py

```diff
@@ -0,0 +1,469 @@
+        value = None
+        lb = sum(w * x.lb if w >= 0 else w * x.ub for w, x in zip(weights, xs))
+        ub = sum(w * x.ub if w >= 0 else w * x.lb for w, x in zip(weights, xs))
+
+        if cmp in ("==", "!="):
+            if lb == rhs == ub:
+                value = BoolVal(True)
+            elif rhs not in range(lb, ub + 1):
+                value = BoolVal(False)
+            if cmp == "!=" and value is not None:
+                value = ~value
+        elif cmp == "<=":
+            if ub <= rhs:
+                value = BoolVal(True)
+            elif lb > rhs:
+                value = BoolVal(False)
+        elif cmp == ">=":
+            if lb >= rhs:
+                value = BoolVal(True)
+            elif ub < rhs:
+                value = BoolVal(False)
+        else:
+            raise NotImplementedError
+
+        # if trivial sat/unsat, return early
+        if value is not None:
+            return [value], []
+
+    terms = []
+    domain_constraints = []
+    for w, x in zip(weights, xs):
+        # the linear may contain Boolean as well as integer variables
+        if isinstance(x, _BoolVarImpl):
+            terms += [(w, x)]
+        else:
+            x_enc, x_cons = _encode_int_var(
+                ivarmap, x, _decide_encoding(x, cmp, encoding)
+            )
+            domain_constraints += x_cons
+            new_terms, k = x_enc.encode_term(
+                w
+            )  # the value of the integer variable as a PB expression `(b_1*c_1) + ... + k`
+            terms += new_terms  # add new terms
+            rhs -= k  # subtract constant from both sides
+
+    # unzip does not work for 0 length iterables
+    if len(terms) == 0:
+        lhs = 0
+    else:
+        # term tuples to two separate lists
+        pb_weights, pb_literals = _unzip(terms)
+
+        # Revert back to `sum` if we happen to have constructed one
+        if all(w == 1 for w in pb_weights):
+            lhs = Operator("sum", pb_literals)
+        else:
+            lhs = Operator("wsum", (pb_weights, pb_literals))
+
+    return [Comparison(cmp, lhs, rhs)], domain_constraints
+
+
+def _encode_comparison(ivarmap, lhs, cmp, rhs, encoding):
+    """Encode integer comparison to PB."""
+    # TODO encode_expr should only use encode linear and check for "comparison" there
+    encoding = _decide_encoding(lhs, cmp, encoding)
+    lhs_enc, domain_constraints = _encode_int_var(ivarmap, lhs, encoding)
+    constraints = lhs_enc.encode_comparison(cmp, rhs)
+    return constraints, domain_constraints
+
+
+def _decide_encoding(x, cmp=None, encoding="auto"):
+    """Decide encoding of `x` via a simple heuristic based on linear comparator and domain size."""
+    if encoding != "auto":
+        return encoding
+    elif _dom_size(x) >= 100:
+        # even slightly large domain sizes suit the binary encoding, as the PB encoding is not expected to be optimized for integer variables!
```

Comment:
I'm not sure what 'not expected to be optimized' means... expected by whom? the user? you as a developer? generally known scientific knowledge?

Reply:
I've clarified comment, let me know if it's more clear now